numpy/typing/tests/data/reveal/char.pyi

```diff
@@ -2,146 +2,209 @@ from typing import Any
+
 assert_type(np.char.rjust(AR_U, 5), npt.NDArray[np.str_])
-assert_type(np.char.rjust(AR_S, [4, 3, 1], fillchar=[b"a", b"b", b"c"]), npt.NDArray[np.bytes_])
+assert_type(np.char.rjust(AR_B, [4, 3, 1], fillchar=[b"a", b"b", b"c"]), npt.NDArray[np.bytes_])
+assert_type(np.char.rjust(AR_S, 5), np.ndarray[np_t._Shape, np.dtypes.StringDType])
 
 assert_type(np.char.lstrip(AR_U), npt.NDArray[np.str_])
-assert_type(np.char.lstrip(AR_S, chars=b"_"), npt.NDArray[np.bytes_])
+assert_type(np.char.lstrip(AR_B, b"_"), npt.NDArray[np.bytes_])
+assert_type(np.char.lstrip(AR_S), np.ndarray[np_t._Shape, np.dtypes.StringDType])
+
 assert_type(np.char.rstrip(AR_U), npt.NDArray[np.str_])
-assert_type(np.char.rstrip(AR_S, chars=b"_"), npt.NDArray[np.bytes_])
+assert_type(np.char.rstrip(AR_B, b"_"), npt.NDArray[np.bytes_])
+assert_type(np.char.rstrip(AR_S), np.ndarray[np_t._Shape, np.dtypes.StringDType])
+
 assert_type(np.char.strip(AR_U), npt.NDArray[np.str_])
-assert_type(np.char.strip(AR_S, chars=b"_"), npt.NDArray[np.bytes_])
+assert_type(np.char.strip(AR_B, b"_"), npt.NDArray[np.bytes_])
+assert_type(np.char.strip(AR_S), np.ndarray[np_t._Shape, np.dtypes.StringDType])
+
+assert_type(np.char.count(AR_U, "a", start=[1, 2, 3]), npt.NDArray[np.int_])
+assert_type(np.char.count(AR_B, [b"a", b"b", b"c"], end=9), npt.NDArray[np.int_])
+assert_type(np.char.count(AR_S, AR_S, start=[1, 2, 3]), npt.NDArray[np.int_])
 
 assert_type(np.char.partition(AR_U, "\n"), npt.NDArray[np.str_])
-assert_type(np.char.partition(AR_S, [b"a", b"b", b"c"]), npt.NDArray[np.bytes_])
+assert_type(np.char.partition(AR_B, [b"a", b"b", b"c"]), npt.NDArray[np.bytes_])
+assert_type(np.char.partition(AR_S, AR_S), np.ndarray[np_t._Shape, np.dtypes.StringDType])
+
 assert_type(np.char.rpartition(AR_U, "\n"), npt.NDArray[np.str_])
-assert_type(np.char.rpartition(AR_S, [b"a", b"b", b"c"]), npt.NDArray[np.bytes_])
+assert_type(np.char.rpartition(AR_B, [b"a", b"b", b"c"]), npt.NDArray[np.bytes_])
+assert_type(np.char.rpartition(AR_S, AR_S), np.ndarray[np_t._Shape, np.dtypes.StringDType])
 
 assert_type(np.char.replace(AR_U, "_", "-"), npt.NDArray[np.str_])
-assert_type(np.char.replace(AR_S, [b"_", b""], [b"a", b"b"]), npt.NDArray[np.bytes_])
+assert_type(np.char.replace(AR_B, [b"_", b""], [b"a", b"b"]), npt.NDArray[np.bytes_])
+assert_type(np.char.replace(AR_S, AR_S, AR_S), np.ndarray[np_t._Shape, np.dtypes.StringDType])
 
 assert_type(np.char.split(AR_U, "_"), npt.NDArray[np.object_])
-assert_type(np.char.split(AR_S, maxsplit=[1, 2, 3]), npt.NDArray[np.object_])
+assert_type(np.char.split(AR_B, maxsplit=[1, 2, 3]), npt.NDArray[np.object_])
+assert_type(np.char.split(AR_S, AR_S), npt.NDArray[np.object_])
+
 assert_type(np.char.rsplit(AR_U, "_"), npt.NDArray[np.object_])
-assert_type(np.char.rsplit(AR_S, maxsplit=[1, 2, 3]), npt.NDArray[np.object_])
+assert_type(np.char.rsplit(AR_B, maxsplit=[1, 2, 3]), npt.NDArray[np.object_])
+assert_type(np.char.rsplit(AR_S, AR_S), npt.NDArray[np.object_])
```

Comment:
This goes back to a decision @lysnikolaou made before 2.0 came out. For functions where the output can't be an array, we're not sure it really makes sense to define them as ufuncs or support them in the future. Rather than making a decision we just decided not to add functions like this to `strings` until we can come up with a better solution. Maybe a padded split with a known output shape and that pads with empty strings?

Reply:
I haven't encountered any use-cases for it myself, so this is all rather new to me. 
When I used it for the first time yesterday, I just assumed that it'd be like a "vectorized `builtins.str.split`", which seemed to work as expected, and I didn't have to look up the docs for it.
So I guess my point is that it might indeed be a good idea to stay close to the `str.split` API 🤷🏻.

And I can see why returning an `object_` array is awkward, but necessary to allow for variable amount of splits. 

Padding seems like a good solution, perhaps with an optional `pad: str = ""` kwarg, so that the new output axis has the size of the max amount of splits. So e.g. `split(["a_b", "a_b_c"], "_")` would then return an array of shape `(2, 3)`, where `out[0, :] == ["a", "b", ""]` (and for `rsplit` you'd then get `["", "a", "b"]` I guess). I suppose that this requires `maxsplit: int`, maybe with `-1` as default to indicate "the longest of all splits".
But you won't be able to use a ufunc anymore in this case; so perhaps a gufunc then?